### PR TITLE
use more granular exit codes when `EasyBuildError` is raised

### PR DESCRIPTION
**Pull request by dagonzalezfo**
#4426 
This PR will provide a basic set of error codes for EB. The included ones can be classified in 2 groups 
-  Eb proper codes: 
     - 3: Missing easyconfig
     - 4: No specific easyblock
     - 5: Failed to import easyblock
     - 6: Failed to  import class from easyblock
     - 7: No recipe for dependency (robot)
     - 8: Os dependency missing
     - 9: Unavailable sources
     - 10: Sanity check folder/file
     - 11: Can not write module at modules path
     - 12: Validate errors (this is more focus on syntax)
- Other error codes: 
    - HTTP: Errors for fetching sources
    - system error codes for system calls
    - LMOD: Errors for loading module

### CODE REVIEW

**Review by lexming — CHANGES_REQUESTED** [review]
I tested this PR and went ahead expanding it to cover more errors. I also created a new `Enum` class to gather the table of exit codes.

Please check PR https://github.com/dagonzalezfo/easybuild-framework/pull/1 and specifically commit https://github.com/dagonzalezfo/easybuild-framework/commit/0d31d34cb5807dd485daf349c847b41f59ce2052 with my proposed changes.

One needed change is that EasyBuild cannot forward any exit codes from HTTP requests, or from external processes, as those can be out of range (3, 128) or have the same value as the custom exit codes of EB, which would be confusing. EasyBuild should always return its own exit codes and HTTP responses or exit codes from external processes should be reported in the logs.

Example behaviour with my PR:
```
$ eb zlib-1.2.13-GCCcore-12.3.0.eb
[...]
== configuring...
  >> running shell command:
        potato --prefix=/home/lexming/easybuild/install/software/zlib/1.2.13-GCCcore-12.3.0
        [started at: 2024-06-11 01:51:20]
        [working dir: /home/lexming/.local/easybuild/build/zlib/1.2.13/GCCcore-12.3.0/zlib-1.2.13]
        [output and state saved to /tmp/eb-8b28i2tz/run-shell-cmd-output/potato-coba1es3]

ERROR: Shell command failed!
    full command              ->  potato --prefix=/home/lexming/easybuild/install/software/zlib/1.2.13-GCCcore-12.3.0
    exit code                 ->  127
    working directory         ->  /home/lexming/.local/easybuild/build/zlib/1.2.13/GCCcore-12.3.0/zlib-1.2.13
    output (stdout + stderr)  ->  /tmp/eb-8b28i2tz/run-shell-cmd-output/potato-coba1es3/out.txt
    called from               ->  'configure_step' function in /home/lexming/.local/venv/eb/lib64/python3.10/site-packages/easybuild/easyblocks/generic/configuremake.py (line 326)

== ... (took < 1 sec)
== FAILED: Installation ended unsuccessfully: shell command 'potato ...' failed with exit code 127 in configure step for zlib-1.2.13-GCCcore-12.3.0.eb (took 0 secs)
== Results of the build can be found in the log file(s) /tmp/eb-8b28i2tz/easybuild-zlib-1.2.13-20240611.015120.QzdZC.log
ERROR: Installation of zlib-1.2.13-GCCcore-12.3.0.eb failed: "shell command 'potato ...' failed with exit code 127 in configure step for zlib-1.2.13-GCCcore-12.3.0.eb"
$ echo $?
15
```

**Comment by lexming** [comment]
@hvelab printing HTTP error codes in warning messages or the logs (which is what that code snippet does) is fine. What is wrong is to use those as exit code for EB as those can easily be out of range  (3, 128). This is why I changed the exit code for EB in that case to well-defined `EasyBuildExit.FAIL_DOWNLOAD`. See a few lines further:

https://github.com/dagonzalezfo/easybuild-framework/blob/0d31d34cb5807dd485daf349c847b41f59ce2052/easybuild/tools/filetools.py#L847

**Comment by hvelab** [comment]
@lexming Great! Then, is there something left to do with this PR from Danilo's help? Please let us know if we can help in any way in your enhancement

**Comment by boegel** [comment]
@hvelab We need to get https://github.com/dagonzalezfo/easybuild-framework/pull/1 merged, which will update this PR.

@lexming I think we can do this without having @dagonzalezfo around though?

**Comment by dagonzalezfo** [comment]
Hi, I will take care of it along this week :)

**Comment by dagonzalezfo** [comment]
Dear @lexming , @boegel, @hvelab ,

I just accept all the changes proposed by Alex, do minimal style fixes and update the PR. 

If it needs futher work, just let me know.

**Review by lexming — DISMISSED** [review]
LGTM

**Comment by dagonzalezfo** [comment]
Dear @lexming, thanks for your help with it

File: easybuild/tools/github.py [hunk 1]
```diff
@@ -1117,8 +1182,10 @@ def _easyconfigs_pr_common(paths, ecs, start_branch=None, pr_branch=None, start_
 
     diff_stat = git_repo.git.diff(cached=True, stat=True)
     if not diff_stat:
-        raise EasyBuildError("No changed files found when comparing to current develop branch. "
-                             "Refused to make empty pull request.")
+        raise EasyBuildError(
+            "No changed files found when comparing to current develop branch. Refused to make empty pull request.",
```
Comment (boegel) [inline easybuild/tools/github.py:1186]: While we're at it: `develop` shouldn't be hardcoded here:
```suggestion
            f"No changed files found when comparing to current {start_branch} branch. "
            "Refused to make empty pull request.",
```

File: easybuild/tools/github.py [hunk 2]
```diff
@@ -1886,7 +1989,10 @@ def new_pr_from_branch(branch_name, title=None, descr=None, pr_target_repo=None,
 
             print_msg('\n'.join(msg), log=_log)
         else:
-            raise EasyBuildError("No changes in '%s' branch compared to current 'develop' branch!", branch_name)
+            raise EasyBuildError(
+                "No changes in '%s' branch compared to current 'develop' branch!", branch_name,
```
Comment (boegel) [inline easybuild/tools/github.py:1993]: shouldn't be hardcoded `develop` here either:
```suggestion
                f"No changes in '{branch_name}' branch compared to current '{pr_target_branch}' branch!",

File: easybuild/tools/github.py [hunk 3]
```diff
@@ -2511,7 +2639,7 @@ def find_easybuild_easyconfig(github_user=None):
     if file_versions:
         fn = sorted(file_versions)[-1][1]
     else:
-        raise EasyBuildError("Couldn't find any EasyBuild easyconfigs")
+        raise EasyBuildError("Couldn't find any EasyBuild easyconfigs", exit_code=EasyBuildExit.MISS_EASYCONFIG)
```
Comment (boegel) [inline easybuild/tools/github.py:2642]:
```suggestion
        raise EasyBuildError("Couldn't find any EasyBuild easyconfigs", exit_code=EasyBuildExit.MISSING_EASYCONFIG)

File: easybuild/tools/options.py [hunk 4]
```diff
@@ -997,15 +1000,22 @@ def _postprocess_optarch(self):
...
                 optarch_dict = {}
                 for compiler, compiler_opt in [p.split(OPTARCH_MAP_CHAR) for p in optarch_parts]:
                     if compiler in optarch_dict:
-                        raise EasyBuildError("The optarch option contains duplicated entries for compiler %s: %s",
-                                             compiler, self.options.optarch)
+                        raise EasyBuildError(
+                            "The optarch option contains duplicated entries for compiler %s: %s",
+                            compiler,
+                            self.options.optarch,
+                            exit_code=EasyBuildExit.OPTION_ERROR
```
Comment (boegel) [inline easybuild/tools/options.py:1017]:
```suggestion
                            compiler, self.options.optarch, exit_code=EasyBuildExit.OPTION_ERROR

File: easybuild/tools/options.py [hunk 5]
```diff
@@ -1017,13 +1027,18 @@ def _postprocess_optarch(self):
...
+            )
 
         reasons = self.options.close_pr_reasons.split(',')
         if any([reason not in VALID_CLOSE_PR_REASONS.keys() for reason in reasons]):
-            raise EasyBuildError("Argument to --close-pr_reasons must be a comma separated list of valid reasons " +
-                                 "among %s" % VALID_CLOSE_PR_REASONS.keys())
+            raise EasyBuildError(
+                "Argument to --close-pr_reasons must be a comma separated list of valid reasons among %s",
+                VALID_CLOSE_PR_REASONS.keys(),
+                exit_code=EasyBuildExit.OPTION_ERROR
```
Comment (boegel) [inline easybuild/tools/options.py:1040]:
```suggestion
                VALID_CLOSE_PR_REASONS.keys(), exit_code=EasyBuildExit.OPTION_ERROR

File: easybuild/tools/options.py [hunk 6]
```diff
@@ -1065,41 +1092,62 @@ def _postprocess_checks(self):
...
+            )
 
         # make sure a GitHub token is available when it's required
         if self.options.upload_test_report:
             if not HAVE_KEYRING:
-                raise EasyBuildError("Python 'keyring' module required for obtaining GitHub token is not available")
+                raise EasyBuildError(
+                    "Python 'keyring' module required for obtaining GitHub token is not available",
+                    exit_code=EasyBuildExit.MISS_EB_DEPENDENCY
```
Comment (boegel) [inline easybuild/tools/options.py:1120]:
```suggestion
                    exit_code=EasyBuildExit.MISSING_EB_DEPENDENCY

File: easybuild/tools/options.py [hunk 7]
```diff
@@ -1065,41 +1092,62 @@ def _postprocess_checks(self):
...
+                    exit_code=EasyBuildExit.FAIL_GITHUB
+                )
 
         # make sure autopep8 is available when it needs to be
         if self.options.dump_autopep8:
             if not HAVE_AUTOPEP8:
-                raise EasyBuildError("Python 'autopep8' module required to reformat dumped easyconfigs as requested")
+                raise EasyBuildError(
+                    "Python 'autopep8' module required to reformat dumped easyconfigs as requested",
+                    exit_code=EasyBuildExit.MISS_EB_DEPENDENCY
```
Comment (boegel) [inline easybuild/tools/options.py:1139]:
```suggestion
                    exit_code=EasyBuildExit.MISSING_EB_DEPENDENCY

File: easybuild/framework/easyblock.py [hunk 8]
```diff
@@ -871,8 +875,11 @@ def obtain_file(self, filename, extension=False, urls=None, download_filename=No
                     return filename
                 else:
                     if not warning_only:
-                        raise EasyBuildError("Couldn't find file %s anywhere, and downloading it is disabled... "
-                                             "Paths attempted (in order): %s ", filename, ', '.join(failedpaths))
+                        raise EasyBuildError(
+                            "Couldn't find file %s anywhere, and downloading it is disabled... "
```
[the remark this answers is not in the document]
Comment (lexming) [reply easybuild/framework/easyblock.py]: sure, done in https://github.com/easybuilders/easybuild-framework/pull/4534/commits/151b95919b490baadec6c021e8e00b18ad08b655

File: easybuild/framework/easyblock.py [hunk 9]
```diff
@@ -4434,6 +4457,7 @@ def ensure_writable_log_dir(log_dir):
         success = True
         summary = 'COMPLETED'
         succ = 'successfully'
+        exit_code = EasyBuildExit.SUCCESS
```
[the remark this answers is not in the document]
Comment (lexming) [reply easybuild/framework/easyblock.py:4466]: done in https://github.com/easybuilders/easybuild-framework/pull/4534/commits/60e06ce360b82120a94a6bca5022fb9486c60b7b

File: easybuild/tools/build_log.py [hunk 10]
```diff
@@ -80,12 +129,14 @@ class EasyBuildError(LoggedException):
...
-    def __init__(self, msg, *args):
+    def __init__(self, msg, *args, **kwargs):
         """Constructor: initialise EasyBuildError instance."""
         if args:
             msg = msg % args
-        LoggedException.__init__(self, msg)
+        LoggedException.__init__(self, msg, **kwargs)
         self.msg = msg
+        if kwargs:
+            self.__dict__.update(kwargs)
```
[the remark this answers is not in the document]
Comment (lexming) [reply easybuild/tools/build_log.py]: this is indeed weird, changed in https://github.com/easybuilders/easybuild-framework/pull/4534/commits/a5e805abcada75cbe369ae4fb019c18fbb4b8954

**Review by lexming — APPROVED** [review]
LGTM

**Comment by boegel** [comment]
Thanks a lot for all the effort on this @dagonzalezfo and @lexming!

**Pull request merged**